aggregator/list.go

```diff
@@ -0,0 +1,340 @@
+			if elem.ReadAndDiscard(alignedStart, l.processAggregatedMetric) {
+				elem.Close()
+				e.Value = nil
+				l.toCollect = append(l.toCollect, e)
+			}
+		default:
+			panic(fmt.Errorf("unknown metric element type: %T", elem))
+		}
+	}
+	l.RUnlock()
+
+	// Flush remaining bytes in the buffer
+	if encoder := l.encoder.Encoder(); len(encoder.Bytes()) > 0 {
+		l.encoder.Reset(l.encoderPool.Get())
+		if err := l.flushFn(encoder); err != nil {
+			l.log.Errorf("flushing metrics error: %v", err)
+		}
+	}
+
+	// Collect tombstoned elements
+	l.Lock()
+	for _, e := range l.toCollect {
+		l.aggregations.Remove(e)
+	}
+	l.Unlock()
+
+	// TODO(xichen): add metrics
+	tickDuration := l.nowFn().Sub(start)
+	if tickDuration < l.flushInterval {
+		// NB(xichen): use a channel here instead of sleeping in case
+		// server needs to close and we don't tick frequently enough
+		select {
+		case <-l.waitForFn(l.flushInterval - tickDuration):
+		case <-l.doneCh:
+		}
+	}
+}
+
+func (l *MetricList) processAggregatedMetric(
+	idPrefix []byte,
+	id metric.ID,
+	idSuffix []byte,
+	timestamp time.Time,
+	value float64,
+	policy policy.Policy,
+) {
+	encoder := l.encoder.Encoder()
+	sizeBefore := encoder.Buffer.Len()
+	if err := l.encodeFn(aggregated.ChunkedMetricWithPolicy{
+		ChunkedMetric: aggregated.ChunkedMetric{
+			ChunkedID: metric.ChunkedID{
+				Prefix: idPrefix,
+				Data:   []byte(id),
+				Suffix: idSuffix,
+			},
+			Timestamp: timestamp,
+			Value:     value,
+		},
+		Policy: policy,
+	}); err != nil {
+		// TODO(xichen): add metrics
+		l.log.WithFields(
+			xlog.NewLogField("idPrefix", string(idPrefix)),
+			xlog.NewLogField("id", id.String()),
+			xlog.NewLogField("idSuffix", string(idSuffix)),
+			xlog.NewLogField("timestamp", timestamp.String()),
+			xlog.NewLogField("value", value),
+			xlog.NewLogField("policy", policy),
+		).Error("encode metric with policy error")
+		encoder.Buffer.Truncate(sizeBefore)
+		// Clear out the encoder error
+		l.encoder.Reset(encoder)
+		return
+	}
+	sizeAfter := encoder.Buffer.Len()
+	// If the buffer size is not big enough, do nothing
+	if sizeAfter < l.maxFlushSize {
+		return
+	}
+	// Otherwise we get a new buffer and copy the bytes exceeding the max
+	// flush size to it, swap the new buffer with the old one, and flush out
+	// the old buffer
+	encoder2 := l.encoderPool.Get()
+	data := encoder.Bytes()
+	encoder2.Buffer.Write(data[sizeBefore:sizeAfter])
+	l.encoder.Reset(encoder2)
+	encoder.Buffer.Truncate(sizeBefore)
+	if err := l.flushFn(encoder); err != nil {
+		l.log.Errorf("flushing metrics error: %v", err)
+	}
+}
+
+type newMetricListFn func(resolution time.Duration, opts Options) *MetricList
+
+// MetricLists contains all the metric lists
+type MetricLists struct {
+	sync.RWMutex
+
+	opts            Options
+	newMetricListFn newMetricListFn
+	closed          bool
+	lists           map[time.Duration]*MetricList
+}
+
+func newMetricLists(opts Options) *MetricLists {
```

Comment:
Yeah possibly, although it's easier to program with a shared lock, since the first thing the aggregator does when adding an entry is to acquire the read lock, if we make it per-list, it'll need to figure out which ones from which lists it needs to RLock, which seems messy.

The critical section within a write lock is really just getting a time value, which should be done instantly, so I think it should be fine. Can optimize further if this becomes the point of contention.